// iframe-shared-storage/dist/__tests__/helpers/client-hub-sandbox.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_IFRAME_ID = void 0;
exports.withClientHubSandbox = withClientHubSandbox;
const jsdom_1 = require("jsdom");
const WINDOW_KEYS = [
    "window",
    "self",
    "document",
    "navigator",
    "addEventListener",
    "removeEventListener",
    "MessageEvent",
    "Event",
    "CustomEvent",
    "HTMLElement",
    "HTMLIFrameElement",
    "HTMLBodyElement",
    "EventTarget",
    "Node",
    "DOMException",
    "performance",
    "getComputedStyle",
];
let currentWindow;
Object.defineProperty(globalThis, "localStorage", {
    configurable: true,
    get() {
        return currentWindow?.localStorage;
    },
});
Object.defineProperty(globalThis, "sessionStorage", {
    configurable: true,
    get() {
        return currentWindow?.sessionStorage;
    },
});
function captureSnapshot() {
    const values = {};
    for (const key of WINDOW_KEYS) {
        values[key] = globalThis[key];
    }
    return { values, currentWindow };
}
function applyWindow(win) {
    for (const key of WINDOW_KEYS) {
        globalThis[key] = win[key];
    }
    currentWindow = win;
}
function restoreWindow(snapshot) {
    for (const key of WINDOW_KEYS) {
        globalThis[key] = snapshot.values[key];
    }
    currentWindow = snapshot.currentWindow;
}
function useWindow(win, fn) {
    const snapshot = captureSnapshot();
    applyWindow(win);
    try {
        return fn();
    }
    finally {
        restoreWindow(snapshot);
    }
}
function linkPostMessage(parent, child) {
    const parentPostMessage = parent.postMessage.bind(parent);
    const childPostMessage = child.postMessage.bind(child);
    const forward = (target, source) => {
        return (data) => {
            setTimeout(() => {
                const event = new target.MessageEvent("message", {
                    data,
                    origin: source.location.origin,
                    source,
                });
                useWindow(target, () => {
                    target.dispatchEvent(event);
                });
            }, 0);
        };
    };
    parent.postMessage = forward(parent, child);
    child.postMessage = forward(child, parent);
    return () => {
        parent.postMessage = parentPostMessage;
        child.postMessage = childPostMessage;
    };
}
function setupLinkedWindows({ iframeId, parentUrl, hubUrl, }) {
    const parentDom = new jsdom_1.JSDOM("<!doctype html><html><body></body></html>", {
        url: parentUrl,
        pretendToBeVisual: true,
    });
    const childDom = new jsdom_1.JSDOM("<!doctype html><html><body></body></html>", {
        url: hubUrl,
        pretendToBeVisual: true,
    });
    const parent = parentDom.window;
    const child = childDom.window;
    Object.defineProperty(child, "parent", {
        configurable: true,
        value: parent,
    });
    Object.defineProperty(parent, "parent", {
        configurable: true,
        value: parent,
    });
    const restorePostMessage = linkPostMessage(parent, child);
    const iframe = parent.document.createElement("iframe");
    iframe.id = iframeId;
    parent.document.body.appendChild(iframe);
    Object.defineProperty(iframe, "contentWindow", {
        configurable: true,
        value: child,
    });
    Object.defineProperty(iframe, "contentDocument", {
        configurable: true,
        value: child.document,
    });
    const { indexedDB, IDBKeyRange } = globalThis;
    Object.assign(parent, { indexedDB, IDBKeyRange });
    Object.assign(child, { indexedDB, IDBKeyRange });
    return {
        parent,
        child,
        cleanup: () => {
            restorePostMessage();
            parent.close();
            child.close();
        },
    };
}
exports.DEFAULT_IFRAME_ID = "integration-frame";
const DEFAULT_PARENT_URL = "http://client.test";
const DEFAULT_HUB_URL = "https://hub.test";
async function withClientHubSandbox(fn, options = {}) {
    const { iframeId = exports.DEFAULT_IFRAME_ID, parentUrl = DEFAULT_PARENT_URL, hubUrl = DEFAULT_HUB_URL, } = options;
    const { parent, child, cleanup } = setupLinkedWindows({
        iframeId,
        parentUrl,
        hubUrl,
    });
    const snapshot = captureSnapshot();
    applyWindow(parent);
    try {
        return await fn({
            iframeId,
            parent,
            child,
            runInParent: (cb) => useWindow(parent, cb),
            runInChild: (cb) => useWindow(child, cb),
        });
    }
    finally {
        restoreWindow(snapshot);
        cleanup();
    }
}
