// iframe-shared-storage/dist/__tests__/client-hub.integration.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = require("../client");
const hub_1 = require("../hub");
const handshake_1 = require("../utils/handshake");
const client_hub_sandbox_1 = require("./helpers/client-hub-sandbox");
describe("client and hub realistic communication", () => {
    it("performs storage RPC roundtrips via postMessage", async () => {
        await (0, client_hub_sandbox_1.withClientHubSandbox)(async ({ iframeId, child, runInChild }) => {
            runInChild(() => (0, hub_1.initHub)());
            const client = (0, client_1.constructClient)({
                iframe: {
                    id: iframeId,
                    iframeReadyTimeoutMs: 500,
                    messagingOptions: {
                    // enableLog: "both",
                    },
                },
            });
            await client.localStorage.setItem("foo", "bar");
            expect(child.localStorage.getItem("foo")).toBe("bar");
            await expect(client.localStorage.getItem("foo")).resolves.toBe("bar");
            await client.localStorage.removeItem("foo");
            expect(child.localStorage.getItem("foo")).toBeNull();
            await client.indexedDBKeyval?.set("key", "value");
            await expect(client.indexedDBKeyval?.get("key")).resolves.toBe("value");
            await client.indexedDBKeyval?.del("key");
            await expect(client.indexedDBKeyval?.get("key")).resolves.toBeUndefined();
        });
    });
    it("rejects calls when the hub iframe never initializes", async () => {
        await (0, client_hub_sandbox_1.withClientHubSandbox)(async ({ iframeId }) => {
            const client = (0, client_1.constructClient)({
                iframe: { id: iframeId, iframeReadyTimeoutMs: 40 },
            });
            await expect(client.localStorage.setItem("delayed", "value")).rejects.toThrow('Iframe storage hub did not initialize within the allowed time before calling method "localStorage.setItem". Waited 40ms.');
        });
    });
    it("retries RPCs that time out before succeeding", async () => {
        await (0, client_hub_sandbox_1.withClientHubSandbox)(async ({ iframeId, runInChild }) => {
            runInChild(() => {
                let shouldDropNextResponse = true;
                const parentWindow = window.parent;
                const originalPostMessage = parentWindow.postMessage.bind(parentWindow);
                parentWindow.postMessage = ((...args) => {
                    const [message] = args;
                    if (shouldDropNextResponse &&
                        message &&
                        typeof message === "object" &&
                        message.sender === "postmsg-rpc/server") {
                        shouldDropNextResponse = false;
                        return;
                    }
                    return originalPostMessage(...args);
                });
                (0, hub_1.initHub)();
            });
            const client = (0, client_1.constructClient)({
                iframe: {
                    id: iframeId,
                    iframeReadyTimeoutMs: 50,
                    methodCallTimeoutMs: 10,
                    methodCallRetries: 1,
                },
            });
            await client.localStorage.setItem("retry-key", "value");
            await expect(client.localStorage.getItem("retry-key")).resolves.toBe("value");
        });
    });
    it("rejects when RPC responses keep timing out even after retries", async () => {
        await (0, client_hub_sandbox_1.withClientHubSandbox)(async ({ iframeId, runInChild }) => {
            runInChild(() => {
                const parentWindow = window.parent;
                const originalPostMessage = parentWindow.postMessage.bind(parentWindow);
                parentWindow.postMessage = ((...args) => {
                    const [message] = args;
                    if (message &&
                        typeof message === "object" &&
                        message.sender === "postmsg-rpc/server") {
                        return;
                    }
                    return originalPostMessage(...args);
                });
                (0, hub_1.initHub)();
            });
            const client = (0, client_1.constructClient)({
                iframe: {
                    id: iframeId,
                    iframeReadyTimeoutMs: 50,
                    methodCallTimeoutMs: 15,
                    methodCallRetries: 2,
                },
            });
            await expect(client.localStorage.getItem("always-timeout")).rejects.toThrow('Iframe storage hub did not respond to method "localStorage.getItem". Waited 15ms.');
        });
    });
    it("rejects calls when the hub never answers RPC messages", async () => {
        await (0, client_hub_sandbox_1.withClientHubSandbox)(async ({ iframeId, runInChild }) => {
            runInChild(() => {
                addEventListener("message", (event) => {
                    if (event.source === window.parent &&
                        (0, handshake_1.isHandshakeMessage)(event.data, handshake_1.HANDSHAKE_REQUEST_TYPE)) {
                        window.parent.postMessage((0, handshake_1.createHandshakeMessage)(handshake_1.HANDSHAKE_RESPONSE_TYPE), event.origin);
                    }
                });
            });
            const client = (0, client_1.constructClient)({
                iframe: {
                    id: iframeId,
                    iframeReadyTimeoutMs: 50,
                    methodCallTimeoutMs: 30,
                },
            });
            await expect(client.localStorage.getItem("never-answers")).rejects.toThrow('Iframe storage hub did not respond to method "localStorage.getItem". Waited 30ms.');
        });
    });
});
