// @cofhe/sdk/dist/node.js
import { createCofheConfigBase, createCofheClientBase } from './chunk-S7OKGLFD.js';
import './chunk-TBLR7NNE.js';
import './chunk-MRCKUMOS.js';
import { TFHE_RS_SAFE_SERIALIZATION_SIZE_LIMIT } from './chunk-4FP4V35O.js';
import { promises } from 'fs';
import { join } from 'path';
import { init_panic_hook, ProvenCompactCiphertextList, CompactPkeCrs, TfheCompactPublicKey } from 'node-tfhe';

var memoryStorage = {};
var createNodeStorage = () => {
  return {
    getItem: async (name) => {
      try {
        const storageDir = join(process.env.HOME || process.env.USERPROFILE || ".", ".cofhesdk");
        await promises.mkdir(storageDir, { recursive: true });
        const filePath = join(storageDir, `${name}.json`);
        const data = await promises.readFile(filePath, "utf8").catch(() => null);
        return data ? JSON.parse(data) : null;
      } catch (e) {
        console.warn("Node.js filesystem modules not available, falling back to memory storage" + e);
        return memoryStorage[name] || null;
      }
    },
    setItem: async (name, value) => {
      try {
        const storageDir = join(process.env.HOME || process.env.USERPROFILE || ".", ".cofhesdk");
        await promises.mkdir(storageDir, { recursive: true });
        const filePath = join(storageDir, `${name}.json`);
        await promises.writeFile(filePath, JSON.stringify(value));
      } catch (e) {
        console.warn("Node.js filesystem modules not available, falling back to memory storage" + e);
        memoryStorage[name] = JSON.stringify(value);
      }
    },
    removeItem: async (name) => {
      try {
        const storageDir = join(process.env.HOME || process.env.USERPROFILE || ".", ".cofhesdk");
        const filePath = join(storageDir, `${name}.json`);
        await promises.unlink(filePath).catch(() => {
        });
      } catch (e) {
        console.warn("Node.js filesystem modules not available, falling back to memory storage" + e);
        delete memoryStorage[name];
      }
    }
  };
};
var tfheInitialized = false;
async function initTfhe() {
  if (tfheInitialized)
    return false;
  await init_panic_hook();
  tfheInitialized = true;
  return true;
}
var fromHexString = (hexString) => {
  const cleanString = hexString.length % 2 === 1 ? `0${hexString}` : hexString;
  const arr = cleanString.replace(/^0x/, "").match(/.{1,2}/g);
  if (!arr)
    return new Uint8Array();
  return new Uint8Array(arr.map((byte) => parseInt(byte, 16)));
};
var _deserializeTfhePublicKey = (buff) => {
  return TfheCompactPublicKey.safe_deserialize(fromHexString(buff), TFHE_RS_SAFE_SERIALIZATION_SIZE_LIMIT);
};
var _deserializeCompactPkeCrs = (buff) => {
  return CompactPkeCrs.safe_deserialize(fromHexString(buff), TFHE_RS_SAFE_SERIALIZATION_SIZE_LIMIT);
};
var tfhePublicKeyDeserializer = (buff) => {
  _deserializeTfhePublicKey(buff);
};
var compactPkeCrsDeserializer = (buff) => {
  _deserializeCompactPkeCrs(buff);
};
var zkBuilderAndCrsGenerator = (fhe, crs) => {
  const fhePublicKey = _deserializeTfhePublicKey(fhe);
  const zkBuilder = ProvenCompactCiphertextList.builder(fhePublicKey);
  const zkCrs = _deserializeCompactPkeCrs(crs);
  return { zkBuilder, zkCrs };
};
function createCofheConfig(config) {
  return createCofheConfigBase({
    environment: "node",
    ...config,
    fheKeyStorage: config.fheKeyStorage === null ? null : config.fheKeyStorage ?? createNodeStorage()
  });
}
function createCofheClient(config) {
  return createCofheClientBase({
    config,
    zkBuilderAndCrsGenerator,
    tfhePublicKeyDeserializer,
    compactPkeCrsDeserializer,
    initTfhe
  });
}

export { createCofheClient, createCofheConfig };
