// @cofhe/sdk/dist/web.js
import { createCofheConfigBase, createCofheClientBase, fheTypeToString } from './chunk-MXND5SVN.js';
import './chunk-TBLR7NNE.js';
import './chunk-NWDKXBIP.js';
import { constructClient } from 'iframe-shared-storage';
import init, { init_panic_hook, TfheCompactPublicKey, CompactPkeCrs, ProvenCompactCiphertextList } from 'tfhe';

var createWebStorage = () => {
  const client = constructClient({
    iframe: {
      src: "https://iframe-shared-storage.vercel.app/hub.html",
      messagingOptions: {
        enableLog: "both"
      },
      iframeReadyTimeoutMs: 3e4,
      // if the iframe is not initied during this interval AND a reuqest is made, such request will throw an error
      methodCallTimeoutMs: 1e4,
      // if a method call is not answered during this interval, the call will throw an error
      methodCallRetries: 3
      // number of retries for a method call if it times out
    }
  });
  const indexedDBKeyval = client.indexedDBKeyval;
  return {
    getItem: async (name) => {
      return await indexedDBKeyval.get(name) ?? null;
    },
    setItem: async (name, value) => {
      await indexedDBKeyval.set(name, value);
    },
    removeItem: async (name) => {
      await indexedDBKeyval.del(name);
    }
  };
};

// web/workerManager.ts
var ZkProveWorkerManager = class {
  worker = null;
  pendingRequests = /* @__PURE__ */ new Map();
  requestCounter = 0;
  workerReady = false;
  initializationPromise = null;
  /**
   * Initialize the worker
   */
  async initializeWorker() {
    if (this.worker && this.workerReady) {
      return;
    }
    if (this.initializationPromise) {
      return this.initializationPromise;
    }
    this.initializationPromise = new Promise((resolve, reject) => {
      try {
        if (typeof Worker === "undefined") {
          reject(new Error("Web Workers not supported"));
          return;
        }
        try {
          this.worker = new Worker(new URL("./zkProve.worker.js", import.meta.url), { type: "module" });
        } catch (error) {
          reject(new Error(`Failed to create worker: ${error}`));
          return;
        }
        this.worker.onmessage = (event) => {
          const { id, type, result, error } = event.data;
          if (type === "ready") {
            this.workerReady = true;
            resolve();
            return;
          }
          const pending = this.pendingRequests.get(id);
          if (!pending) {
            console.warn("[Worker Manager] Received response for unknown request:", id);
            return;
          }
          clearTimeout(pending.timeoutId);
          this.pendingRequests.delete(id);
          if (type === "success" && result) {
            pending.resolve(new Uint8Array(result));
          } else if (type === "error") {
            pending.reject(new Error(error || "Worker error"));
          } else {
            pending.reject(new Error("Invalid response from worker"));
          }
        };
        this.worker.onerror = (error) => {
          console.error("[Worker Manager] Worker error event:", error);
          console.error("[Worker Manager] Error message:", error.message);
          console.error("[Worker Manager] Error filename:", error.filename);
          console.error("[Worker Manager] Error lineno:", error.lineno);
          if (!this.workerReady) {
            reject(new Error(`Worker failed to initialize: ${error.message || "Unknown error"}`));
          }
          this.pendingRequests.forEach(({ reject: reject2, timeoutId }) => {
            clearTimeout(timeoutId);
            reject2(new Error("Worker encountered an error"));
          });
          this.pendingRequests.clear();
        };
        setTimeout(() => {
          if (!this.workerReady) {
            reject(new Error("Worker initialization timeout"));
          }
        }, 5e3);
      } catch (error) {
        reject(error);
      }
    });
    return this.initializationPromise;
  }
  /**
   * Submit a proof generation request to the worker
   */
  async submitProof(fheKeyHex, crsHex, items, metadata) {
    await this.initializeWorker();
    const id = `zkprove-${Date.now()}-${this.requestCounter++}`;
    return new Promise((resolve, reject) => {
      const timeoutId = setTimeout(() => {
        this.pendingRequests.delete(id);
        reject(new Error("Worker request timeout (30s)"));
      }, 3e4);
      this.pendingRequests.set(id, { resolve, reject, timeoutId });
      const message = {
        id,
        type: "zkProve",
        fheKeyHex,
        crsHex,
        items,
        metadata: Array.from(metadata)
      };
      this.worker.postMessage(message);
    });
  }
  /**
   * Terminate the worker and clean up
   */
  terminate() {
    if (this.worker) {
      this.worker.terminate();
      this.worker = null;
      this.workerReady = false;
      this.initializationPromise = null;
    }
    this.pendingRequests.forEach(({ reject, timeoutId }) => {
      clearTimeout(timeoutId);
      reject(new Error("Worker terminated"));
    });
    this.pendingRequests.clear();
  }
  /**
   * Check if worker is available
   */
  isAvailable() {
    return typeof Worker !== "undefined";
  }
};
var workerManager = null;
function getWorkerManager() {
  if (!workerManager) {
    workerManager = new ZkProveWorkerManager();
  }
  return workerManager;
}
function terminateWorker() {
  if (workerManager) {
    workerManager.terminate();
    workerManager = null;
  }
}
function areWorkersAvailable() {
  return typeof Worker !== "undefined";
}
var tfheInitialized = false;
async function initTfhe() {
  if (tfheInitialized)
    return false;
  await init();
  await init_panic_hook();
  tfheInitialized = true;
  return true;
}
var fromHexString = (hexString) => {
  const cleanString = hexString.length % 2 === 1 ? `0${hexString}` : hexString;
  const arr = cleanString.replace(/^0x/, "").match(/.{1,2}/g);
  if (!arr)
    return new Uint8Array();
  return new Uint8Array(arr.map((byte) => parseInt(byte, 16)));
};
var tfhePublicKeyDeserializer = (buff) => {
  TfheCompactPublicKey.deserialize(fromHexString(buff));
};
var compactPkeCrsDeserializer = (buff) => {
  CompactPkeCrs.deserialize(fromHexString(buff));
};
var zkBuilderAndCrsGenerator = (fhe, crs) => {
  const fhePublicKey = TfheCompactPublicKey.deserialize(fromHexString(fhe));
  const zkBuilder = ProvenCompactCiphertextList.builder(fhePublicKey);
  const zkCrs = CompactPkeCrs.deserialize(fromHexString(crs));
  return { zkBuilder, zkCrs };
};
async function zkProveWithWorker(fheKeyHex, crsHex, items, metadata) {
  const serializedItems = items.map((item) => ({
    utype: fheTypeToString(item.utype),
    data: typeof item.data === "bigint" ? item.data.toString() : item.data
  }));
  const workerManager2 = getWorkerManager();
  return await workerManager2.submitProof(fheKeyHex, crsHex, serializedItems, metadata);
}
function createCofheConfig(config) {
  return createCofheConfigBase({
    environment: "web",
    ...config,
    fheKeyStorage: config.fheKeyStorage === null ? null : config.fheKeyStorage ?? createWebStorage()
  });
}
function createCofheClient(config) {
  return createCofheClientBase({
    config,
    zkBuilderAndCrsGenerator,
    tfhePublicKeyDeserializer,
    compactPkeCrsDeserializer,
    initTfhe,
    // Always provide the worker function if available - config.useWorkers controls usage
    // areWorkersAvailable will return true if the Worker API is available and false in Node.js
    zkProveWorkerFn: areWorkersAvailable() ? zkProveWithWorker : void 0
  });
}
function createCofheClientWithCustomWorker(config, customZkProveWorkerFn) {
  return createCofheClientBase({
    config,
    zkBuilderAndCrsGenerator,
    tfhePublicKeyDeserializer,
    compactPkeCrsDeserializer,
    initTfhe,
    zkProveWorkerFn: customZkProveWorkerFn
  });
}

export { areWorkersAvailable, createCofheClient, createCofheClientWithCustomWorker, createCofheConfig, terminateWorker };
