// @cofhe/sdk/dist/chains.cjs
'use strict';

var zod = require('zod');

// chains/types.ts
var EnvironmentSchema = zod.z.enum(["MOCK", "TESTNET", "MAINNET"]);
var CofheChainSchema = zod.z.object({
  /** Chain ID */
  id: zod.z.number().int().positive(),
  /** Human-readable chain name */
  name: zod.z.string().min(1),
  /** Network identifier */
  network: zod.z.string().min(1),
  /** coFhe service URL */
  coFheUrl: zod.z.url(),
  /** Verifier service URL */
  verifierUrl: zod.z.url(),
  /** Threshold network service URL */
  thresholdNetworkUrl: zod.z.url(),
  /** Environment type */
  environment: EnvironmentSchema
});
function defineChain(chainConfig) {
  const result = CofheChainSchema.safeParse(chainConfig);
  if (!result.success) {
    throw new Error(`Invalid chain configuration: ${zod.z.prettifyError(result.error)}`, { cause: result.error });
  }
  return result.data;
}

// chains/chains/sepolia.ts
var sepolia = defineChain({
  id: 11155111,
  name: "Sepolia",
  network: "sepolia",
  coFheUrl: "https://testnet-cofhe.fhenix.zone",
  verifierUrl: "https://testnet-cofhe-vrf.fhenix.zone",
  thresholdNetworkUrl: "https://testnet-cofhe-tn.fhenix.zone",
  environment: "TESTNET"
});

// chains/chains/arbSepolia.ts
var arbSepolia = defineChain({
  id: 421614,
  name: "Arbitrum Sepolia",
  network: "arb-sepolia",
  coFheUrl: "https://testnet-cofhe.fhenix.zone",
  verifierUrl: "https://testnet-cofhe-vrf.fhenix.zone",
  thresholdNetworkUrl: "https://testnet-cofhe-tn.fhenix.zone",
  environment: "TESTNET"
});

// chains/chains/baseSepolia.ts
var baseSepolia = defineChain({
  id: 84532,
  name: "Base Sepolia",
  network: "base-sepolia",
  coFheUrl: "https://testnet-cofhe.fhenix.zone",
  verifierUrl: "https://testnet-cofhe-vrf.fhenix.zone",
  thresholdNetworkUrl: "https://testnet-cofhe-tn.fhenix.zone",
  environment: "TESTNET"
});

// chains/chains/hardhat.ts
var hardhat = defineChain({
  id: 31337,
  name: "Hardhat",
  network: "localhost",
  // These are unused in the mock environment
  coFheUrl: "http://127.0.0.1:8448",
  verifierUrl: "http://127.0.0.1:3001",
  thresholdNetworkUrl: "http://127.0.0.1:3000",
  environment: "MOCK"
});

// chains/chains/localcofhe.ts
var localcofhe = defineChain({
  id: 420105,
  name: "Local Cofhe",
  network: "localhost",
  coFheUrl: "http://127.0.0.1:9448",
  verifierUrl: "http://127.0.0.1:3001",
  thresholdNetworkUrl: "http://127.0.0.1:3000",
  environment: "TESTNET"
});

// chains/index.ts
var chains = {
  sepolia,
  arbSepolia,
  baseSepolia,
  hardhat,
  localcofhe
};
var getChainById = (chainId) => {
  return Object.values(chains).find((chain) => chain.id === chainId);
};
var getChainByName = (name) => {
  return Object.values(chains).find(
    (chain) => chain.name.toLowerCase() === name.toLowerCase() || chain.network.toLowerCase() === name.toLowerCase()
  );
};

exports.arbSepolia = arbSepolia;
exports.baseSepolia = baseSepolia;
exports.chains = chains;
exports.getChainById = getChainById;
exports.getChainByName = getChainByName;
exports.hardhat = hardhat;
exports.localcofhe = localcofhe;
exports.sepolia = sepolia;
