// iframe-shared-storage/dist/browser.js
"use strict";
var IframeStorage = (() => {
  var __create = Object.create;
  var __defProp = Object.defineProperty;
  var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
  var __getOwnPropNames = Object.getOwnPropertyNames;
  var __getProtoOf = Object.getPrototypeOf;
  var __hasOwnProp = Object.prototype.hasOwnProperty;
  var __esm = (fn, res) => function __init() {
    return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
  };
  var __commonJS = (cb, mod) => function __require() {
    return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
  };
  var __export = (target, all) => {
    for (var name in all)
      __defProp(target, name, { get: all[name], enumerable: true });
  };
  var __copyProps = (to, from, except, desc) => {
    if (from && typeof from === "object" || typeof from === "function") {
      for (let key of __getOwnPropNames(from))
        if (!__hasOwnProp.call(to, key) && key !== except)
          __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
    }
    return to;
  };
  var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
    // If the importer is in node compatibility mode or this is not an ESM
    // file that has been converted to a CommonJS file using a Babel-
    // compatible transform (i.e. "__esModule" has not been set), then set
    // "default" to the CommonJS "module.exports" for node compatibility.
    isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
    mod
  ));
  var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

  // node_modules/shortid/lib/random/random-from-seed.js
  var require_random_from_seed = __commonJS({
    "node_modules/shortid/lib/random/random-from-seed.js"(exports, module) {
      "use strict";
      var seed = 1;
      function getNextValue() {
        seed = (seed * 9301 + 49297) % 233280;
        return seed / 233280;
      }
      function setSeed(_seed_) {
        seed = _seed_;
      }
      module.exports = {
        nextValue: getNextValue,
        seed: setSeed
      };
    }
  });

  // node_modules/shortid/lib/alphabet.js
  var require_alphabet = __commonJS({
    "node_modules/shortid/lib/alphabet.js"(exports, module) {
      "use strict";
      var randomFromSeed = require_random_from_seed();
      var ORIGINAL = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ_-";
      var alphabet;
      var previousSeed;
      var shuffled;
      function reset() {
        shuffled = false;
      }
      function setCharacters(_alphabet_) {
        if (!_alphabet_) {
          if (alphabet !== ORIGINAL) {
            alphabet = ORIGINAL;
            reset();
          }
          return;
        }
        if (_alphabet_ === alphabet) {
          return;
        }
        if (_alphabet_.length !== ORIGINAL.length) {
          throw new Error("Custom alphabet for shortid must be " + ORIGINAL.length + " unique characters. You submitted " + _alphabet_.length + " characters: " + _alphabet_);
        }
        var unique = _alphabet_.split("").filter(function(item, ind, arr) {
          return ind !== arr.lastIndexOf(item);
        });
        if (unique.length) {
          throw new Error("Custom alphabet for shortid must be " + ORIGINAL.length + " unique characters. These characters were not unique: " + unique.join(", "));
        }
        alphabet = _alphabet_;
        reset();
      }
      function characters(_alphabet_) {
        setCharacters(_alphabet_);
        return alphabet;
      }
      function setSeed(seed) {
        randomFromSeed.seed(seed);
        if (previousSeed !== seed) {
          reset();
          previousSeed = seed;
        }
      }
      function shuffle() {
        if (!alphabet) {
          setCharacters(ORIGINAL);
        }
        var sourceArray = alphabet.split("");
        var targetArray = [];
        var r = randomFromSeed.nextValue();
        var characterIndex;
        while (sourceArray.length > 0) {
          r = randomFromSeed.nextValue();
          characterIndex = Math.floor(r * sourceArray.length);
          targetArray.push(sourceArray.splice(characterIndex, 1)[0]);
        }
        return targetArray.join("");
      }
      function getShuffled() {
        if (shuffled) {
          return shuffled;
        }
        shuffled = shuffle();
        return shuffled;
      }
      function lookup(index) {
        var alphabetShuffled = getShuffled();
        return alphabetShuffled[index];
      }
      function get2() {
        return alphabet || ORIGINAL;
      }
      module.exports = {
        get: get2,
        characters,
        seed: setSeed,
        lookup,
        shuffled: getShuffled
      };
    }
  });

  // node_modules/shortid/lib/random/random-byte-browser.js
  var require_random_byte_browser = __commonJS({
    "node_modules/shortid/lib/random/random-byte-browser.js"(exports, module) {
      "use strict";
      var crypto2 = typeof window === "object" && (window.crypto || window.msCrypto);
      var randomByte;
      if (!crypto2 || !crypto2.getRandomValues) {
        randomByte = function(size) {
          var bytes = [];
          for (var i = 0; i < size; i++) {
            bytes.push(Math.floor(Math.random() * 256));
          }
          return bytes;
        };
      } else {
        randomByte = function(size) {
          return crypto2.getRandomValues(new Uint8Array(size));
        };
      }
      module.exports = randomByte;
    }
  });

  // node_modules/nanoid/url-alphabet/index.js
  var urlAlphabet;
  var init_url_alphabet = __esm({
    "node_modules/nanoid/url-alphabet/index.js"() {
      urlAlphabet = "useandom-26T198340PX75pxJACKVERYMINDBUSHWOLF_GQZbfghjklqvwyzrict";
    }
  });

  // node_modules/nanoid/index.browser.js
  var index_browser_exports = {};
  __export(index_browser_exports, {
    customAlphabet: () => customAlphabet,
    customRandom: () => customRandom,
    nanoid: () => nanoid,
    random: () => random,
    urlAlphabet: () => urlAlphabet
  });
  var random, customRandom, customAlphabet, nanoid;
  var init_index_browser = __esm({
    "node_modules/nanoid/index.browser.js"() {
      init_url_alphabet();
      random = (bytes) => crypto.getRandomValues(new Uint8Array(bytes));
      customRandom = (alphabet, defaultSize, getRandom) => {
        let mask = (2 << Math.log(alphabet.length - 1) / Math.LN2) - 1;
        let step = -~(1.6 * mask * defaultSize / alphabet.length);
        return (size = defaultSize) => {
          let id = "";
          while (true) {
            let bytes = getRandom(step);
            let j = step | 0;
            while (j--) {
              id += alphabet[bytes[j] & mask] || "";
              if (id.length === size) return id;
            }
          }
        };
      };
      customAlphabet = (alphabet, size = 21) => customRandom(alphabet, size, random);
      nanoid = (size = 21) => crypto.getRandomValues(new Uint8Array(size)).reduce((id, byte) => {
        byte &= 63;
        if (byte < 36) {
          id += byte.toString(36);
        } else if (byte < 62) {
          id += (byte - 26).toString(36).toUpperCase();
        } else if (byte > 62) {
          id += "-";
        } else {
          id += "_";
        }
        return id;
      }, "");
    }
  });

  // node_modules/shortid/lib/generate.js
  var require_generate = __commonJS({
    "node_modules/shortid/lib/generate.js"(exports, module) {
      "use strict";
      var alphabet = require_alphabet();
      var random2 = require_random_byte_browser();
      var customRandom2 = (init_index_browser(), __toCommonJS(index_browser_exports)).customRandom;
      function generate(number) {
        var loopCounter = 0;
        var done;
        var str = "";
        while (!done) {
          str = str + customRandom2(alphabet.get(), 1, random2)();
          done = number < Math.pow(16, loopCounter + 1);
          loopCounter++;
        }
        return str;
      }
      module.exports = generate;
    }
  });

  // node_modules/shortid/lib/build.js
  var require_build = __commonJS({
    "node_modules/shortid/lib/build.js"(exports, module) {
      "use strict";
      var generate = require_generate();
      var alphabet = require_alphabet();
      var REDUCE_TIME = 1567752802062;
      var version = 7;
      var counter;
      var previousSeconds;
      function build(clusterWorkerId) {
        var str = "";
        var seconds = Math.floor((Date.now() - REDUCE_TIME) * 1e-3);
        if (seconds === previousSeconds) {
          counter++;
        } else {
          counter = 0;
          previousSeconds = seconds;
        }
        str = str + generate(version);
        str = str + generate(clusterWorkerId);
        if (counter > 0) {
          str = str + generate(counter);
        }
        str = str + generate(seconds);
        return str;
      }
      module.exports = build;
    }
  });

  // node_modules/shortid/lib/is-valid.js
  var require_is_valid = __commonJS({
    "node_modules/shortid/lib/is-valid.js"(exports, module) {
      "use strict";
      var alphabet = require_alphabet();
      function isShortId(id) {
        if (!id || typeof id !== "string" || id.length < 6) {
          return false;
        }
        var nonAlphabetic = new RegExp("[^" + alphabet.get().replace(/[|\\{}()[\]^$+*?.-]/g, "\\$&") + "]");
        return !nonAlphabetic.test(id);
      }
      module.exports = isShortId;
    }
  });

  // node_modules/shortid/lib/util/cluster-worker-id-browser.js
  var require_cluster_worker_id_browser = __commonJS({
    "node_modules/shortid/lib/util/cluster-worker-id-browser.js"(exports, module) {
      "use strict";
      module.exports = 0;
    }
  });

  // node_modules/shortid/lib/index.js
  var require_lib = __commonJS({
    "node_modules/shortid/lib/index.js"(exports, module) {
      "use strict";
      var alphabet = require_alphabet();
      var build = require_build();
      var isValid = require_is_valid();
      var clusterWorkerId = require_cluster_worker_id_browser() || 0;
      function seed(seedValue) {
        alphabet.seed(seedValue);
        return module.exports;
      }
      function worker(workerId) {
        clusterWorkerId = workerId;
        return module.exports;
      }
      function characters(newCharacters) {
        if (newCharacters !== void 0) {
          alphabet.characters(newCharacters);
        }
        return alphabet.shuffled();
      }
      function generate() {
        return build(clusterWorkerId);
      }
      module.exports = generate;
      module.exports.generate = generate;
      module.exports.seed = seed;
      module.exports.worker = worker;
      module.exports.characters = characters;
      module.exports.isValid = isValid;
    }
  });

  // node_modules/shortid/index.js
  var require_shortid = __commonJS({
    "node_modules/shortid/index.js"(exports, module) {
      "use strict";
      module.exports = require_lib();
    }
  });

  // node_modules/postmsg-rpc/lib/client.js
  var require_client = __commonJS({
    "node_modules/postmsg-rpc/lib/client.js"(exports) {
      "use strict";
      Object.defineProperty(exports, "__esModule", {
        value: true
      });
      exports.default = caller2;
      exports.call = call;
      var _shortid = require_shortid();
      var _shortid2 = _interopRequireDefault(_shortid);
      function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : { default: obj };
      }
      function caller2(funcName, opts) {
        opts = opts || {};
        var addListener = opts.addListener || window.addEventListener;
        var removeListener = opts.removeListener || window.removeEventListener;
        var postMessage = opts.postMessage || window.postMessage;
        var targetOrigin = opts.targetOrigin || "*";
        var getMessageData = opts.getMessageData || function(event) {
          return event.data;
        };
        return function() {
          var msg = {
            sender: "postmsg-rpc/client",
            id: (0, _shortid2.default)(),
            func: funcName,
            args: Array.from(arguments)
          };
          var cancel = void 0;
          var response = new Promise(function(resolve, reject) {
            var handler = function handler2() {
              var data = getMessageData.apply(null, arguments);
              if (!data) return;
              if (data.sender !== "postmsg-rpc/server" || data.id !== msg.id) return;
              removeListener("message", handler2);
              if (data.err) {
                var err = new Error("Unexpected error calling " + funcName);
                Object.assign(err, data.err);
                return reject(err);
              }
              resolve(data.res);
            };
            cancel = function cancel2() {
              removeListener("message", handler);
              var err = new Error("Canceled call to " + funcName);
              err.isCanceled = true;
              reject(err);
            };
            addListener("message", handler);
            postMessage(msg, targetOrigin);
          });
          response.cancel = function() {
            return cancel();
          };
          return response;
        };
      }
      function call(funcName) {
        return caller2(funcName).apply(null, [].slice.call(arguments, 1));
      }
    }
  });

  // node_modules/postmsg-rpc/lib/server.js
  var require_server = __commonJS({
    "node_modules/postmsg-rpc/lib/server.js"(exports) {
      "use strict";
      Object.defineProperty(exports, "__esModule", {
        value: true
      });
      exports.default = expose2;
      function expose2(funcName, func, opts) {
        opts = opts || {};
        var addListener = opts.addListener || window.addEventListener;
        var removeListener = opts.removeListener || window.removeEventListener;
        var postMessage = opts.postMessage || window.postMessage;
        var targetOrigin = opts.targetOrigin || "*";
        var getMessageData = opts.getMessageData || function(event) {
          return event.data;
        };
        var isCallback = opts.isCallback || false;
        var handler = function handler2() {
          var data = getMessageData.apply(null, arguments);
          if (!data) return;
          if (data.sender !== "postmsg-rpc/client" || data.func !== funcName) return;
          var msg = { sender: "postmsg-rpc/server", id: data.id };
          var onSuccess = function onSuccess2(res2) {
            msg.res = res2;
            postMessage(msg, targetOrigin);
          };
          var onError = function onError2(err) {
            msg.err = Object.assign({ message: err.message }, err.output && err.output.payload);
            if (true) {
              msg.err.stack = msg.err.stack || err.stack;
            }
            postMessage(msg, targetOrigin);
          };
          if (isCallback) {
            func.apply(null, data.args.concat(function(err, res2) {
              if (err) return onError(err);
              onSuccess(res2);
            }));
          } else {
            var res = func.apply(null, data.args);
            Promise.resolve(res).then(onSuccess).catch(onError);
          }
        };
        addListener("message", handler);
        return { close: function close() {
          return removeListener("message", handler);
        } };
      }
    }
  });

  // node_modules/postmsg-rpc/lib/index.js
  var require_lib2 = __commonJS({
    "node_modules/postmsg-rpc/lib/index.js"(exports) {
      "use strict";
      Object.defineProperty(exports, "__esModule", {
        value: true
      });
      var _client = require_client();
      Object.defineProperty(exports, "caller", {
        enumerable: true,
        get: function get2() {
          return _interopRequireDefault(_client).default;
        }
      });
      Object.defineProperty(exports, "call", {
        enumerable: true,
        get: function get2() {
          return _client.call;
        }
      });
      var _server = require_server();
      Object.defineProperty(exports, "expose", {
        enumerable: true,
        get: function get2() {
          return _interopRequireDefault(_server).default;
        }
      });
      function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : { default: obj };
      }
    }
  });

  // src/index.ts
  var index_exports = {};
  __export(index_exports, {
    constructClient: () => constructClient,
    initHub: () => initHub
  });

  // src/hub.ts
  var import_postmsg_rpc = __toESM(require_lib2());

  // node_modules/idb-keyval/dist/index.js
  function promisifyRequest(request) {
    return new Promise((resolve, reject) => {
      request.oncomplete = request.onsuccess = () => resolve(request.result);
      request.onabort = request.onerror = () => reject(request.error);
    });
  }
  function createStore(dbName, storeName) {
    let dbp;
    const getDB = () => {
      if (dbp)
        return dbp;
      const request = indexedDB.open(dbName);
      request.onupgradeneeded = () => request.result.createObjectStore(storeName);
      dbp = promisifyRequest(request);
      dbp.then((db) => {
        db.onclose = () => dbp = void 0;
      }, () => {
      });
      return dbp;
    };
    return (txMode, callback) => getDB().then((db) => callback(db.transaction(storeName, txMode).objectStore(storeName)));
  }
  var defaultGetStoreFunc;
  function defaultGetStore() {
    if (!defaultGetStoreFunc) {
      defaultGetStoreFunc = createStore("keyval-store", "keyval");
    }
    return defaultGetStoreFunc;
  }
  function get(key, customStore = defaultGetStore()) {
    return customStore("readonly", (store) => promisifyRequest(store.get(key)));
  }
  function set(key, value, customStore = defaultGetStore()) {
    return customStore("readwrite", (store) => {
      store.put(value, key);
      return promisifyRequest(store.transaction);
    });
  }
  function del(key, customStore = defaultGetStore()) {
    return customStore("readwrite", (store) => {
      store.delete(key);
      return promisifyRequest(store.transaction);
    });
  }

  // src/utils/handshake.ts
  var HANDSHAKE_NAMESPACE = "iframe-storage:hub-ready";
  var HANDSHAKE_REQUEST_TYPE = `${HANDSHAKE_NAMESPACE}:request`;
  var HANDSHAKE_RESPONSE_TYPE = `${HANDSHAKE_NAMESPACE}:response`;
  function createHandshakeMessage(type) {
    return {
      __iframeStorageHandshake: true,
      type
    };
  }
  function isHandshakeMessage(data, type) {
    if (typeof data !== "object" || data === null || data.__iframeStorageHandshake !== true) {
      return false;
    }
    if (!type) {
      return true;
    }
    return data.type === type;
  }

  // src/utils/log.ts
  function logIfEnabled(options, domain, action, ...info) {
    if (!options?.enableLog) return;
    if (options.enableLog !== "both" && options.enableLog !== domain) return;
    console.log(`[iframe-storage:${domain}:${action}]`, ...info);
  }

  // src/hub.ts
  function initHub() {
    if (!window?.parent)
      throw new Error("Hub must be run inside an iframe with a parent window.");
    addEventListener("message", (event) => {
      if (event.source === window.parent && isHandshakeMessage(event.data, HANDSHAKE_REQUEST_TYPE)) {
        window.parent.postMessage(
          createHandshakeMessage(HANDSHAKE_RESPONSE_TYPE),
          event.origin
        );
        return;
      }
      const messagingOptions = event.data?.args ? event.data.args[event.data.args.length - 1] : void 0;
      if (messagingOptions?.enableLog !== "hub" && messagingOptions?.enableLog !== "both")
        return;
      const method = event.data?.func || "unknown_method";
      const id = event.data?.id || "unknown_id";
      logIfEnabled(
        messagingOptions,
        "hub",
        `${id}:${method}`,
        "received_message",
        JSON.stringify(event.data).substring(0, 200)
      );
    });
    const localStorageMethods = {
      ["localStorage.setItem" /* LocalStorage_SetItem */]: (key, value) => localStorage.setItem(key, value),
      ["localStorage.getItem" /* LocalStorage_GetItem */]: (key) => localStorage.getItem(key),
      ["localStorage.removeItem" /* LocalStorage_RemoveItem */]: (key) => localStorage.removeItem(key),
      ["localStorage.clear" /* LocalStorage_Clear */]: () => localStorage.clear(),
      ["localStorage.key" /* LocalStorage_Key */]: (index) => localStorage.key(index)
    };
    const indexDBKeyvalMethods = {
      ["indexDBKeyval.set" /* indexDBKeyval_Set */]: (key, value) => set(key, value),
      ["indexDBKeyval.get" /* indexDBKeyval_Get */]: (key) => get(key),
      ["indexDBKeyval.del" /* indexDBKeyval_Del */]: (key) => del(key)
      // Note: idb-keyval does not have clear and key methods, so we skip them
    };
    const hubService = {
      ...localStorageMethods,
      ...indexDBKeyvalMethods
    };
    for (const [methodName, methodImpl] of Object.entries(hubService)) {
      (0, import_postmsg_rpc.expose)(
        methodName,
        async (...args) => {
          const maybeOptions = args.length ? args[args.length - 1] : void 0;
          const loggedArgs = maybeOptions ? args.slice(0, args.length - 1) : args;
          logIfEnabled(
            maybeOptions,
            "hub",
            methodName,
            "before_call",
            JSON.stringify(loggedArgs).substring(0, 200)
          );
          const result = await methodImpl(...args);
          logIfEnabled(
            maybeOptions,
            "hub",
            methodName,
            "after_call",
            String(result).substring(0, 200)
          );
          return result;
        },
        {
          postMessage: window.parent.postMessage.bind(window.parent)
        }
      );
    }
  }

  // src/client.ts
  var import_postmsg_rpc2 = __toESM(require_lib2());

  // src/utils/await-with-timeout.ts
  async function awaitWithTimeout(promise, timeoutMs, onTimeout) {
    let timeoutId;
    const timeoutPromise = new Promise((resolve, reject) => {
      timeoutId = setTimeout(() => {
        try {
          if (!onTimeout) {
            reject(new Error("Operation timed out."));
            return;
          }
          const result = onTimeout();
          resolve(result);
        } catch (error) {
          reject(error);
        }
      }, timeoutMs);
    });
    return Promise.race([promise, timeoutPromise]).finally(() => {
      if (typeof timeoutId !== "undefined") {
        clearTimeout(timeoutId);
      }
    });
  }

  // src/client.ts
  var DEFAULT_IFRAME_ID = "iframe-storage-hub";
  var DEFAULT_INITIALIZATION_TIMEOUT_MS = 1e3;
  var DEFAULT_METHOD_CALL_TIMEOUT_MS = 1e3;
  var DEFAULT_METHOD_CALL_RETRIES = 0;
  var MethodCallTimeoutError = class extends Error {
    constructor(method, timeoutMs) {
      super(
        `Iframe storage hub did not respond to method "${method}". Waited ${timeoutMs}ms.`
      );
      this.method = method;
      this.timeoutMs = timeoutMs;
      this.name = "IframeStorageMethodTimeoutError";
    }
  };
  function constructClient({ iframe }) {
    const {
      messagingOptions,
      iframeReadyTimeoutMs = DEFAULT_INITIALIZATION_TIMEOUT_MS,
      methodCallTimeoutMs = DEFAULT_METHOD_CALL_TIMEOUT_MS,
      methodCallRetries = DEFAULT_METHOD_CALL_RETRIES
    } = iframe;
    const {
      postMessage,
      ready: iframeReady,
      getWaitDurationMs: getIframeWaitDurationMs,
      isReady: isIframeReady
    } = createIframePostMessage(iframe);
    const callerOptions = { postMessage };
    const callerWithOptions = async (method, ...args) => {
      logIfEnabled(
        messagingOptions,
        "client",
        method,
        "before-postmessage",
        JSON.stringify(args).substring(0, 200)
      );
      const performCall = async () => {
        if (!isIframeReady()) {
          const elapsed = getIframeWaitDurationMs();
          const remaining = iframeReadyTimeoutMs - elapsed;
          if (remaining <= 0) {
            throw new Error(
              `Iframe storage hub did not initialize within the allowed time before calling method "${method}". Waited ${elapsed}ms.`
            );
          }
          await awaitWithTimeout(iframeReady, remaining, () => {
            throw new Error(
              `Iframe storage hub did not initialize within the allowed time before calling method "${method}". Waited ${iframeReadyTimeoutMs}ms.`
            );
          });
        }
        const rpcPromise = (0, import_postmsg_rpc2.caller)(method, callerOptions)(
          ...args,
          messagingOptions
        );
        return await awaitWithTimeout(rpcPromise, methodCallTimeoutMs, () => {
          throw new MethodCallTimeoutError(method, methodCallTimeoutMs);
        });
      };
      let attempt = 0;
      while (attempt <= methodCallRetries) {
        try {
          return await performCall();
        } catch (error) {
          const isTimeoutError = error instanceof MethodCallTimeoutError || error instanceof Error && error.name === "IframeStorageMethodTimeoutError";
          const remainingAttempts = methodCallRetries - attempt;
          if (!isTimeoutError || remainingAttempts <= 0) {
            throw error;
          }
          logIfEnabled(
            messagingOptions,
            "client",
            method,
            "retry-after-timeout",
            `attempt=${attempt + 1}`,
            `remaining_attempts=${remainingAttempts - 1}`
          );
          attempt += 1;
        }
      }
      throw new Error("Unreachable RPC retry loop exit.");
    };
    addEventListener("message", (event) => {
      const sender = event.data.sender;
      if ("postmsg-rpc/server" !== sender) return;
      if (messagingOptions?.enableLog !== "client" && messagingOptions?.enableLog !== "both")
        return;
      const id = event.data?.id || "unknown_id";
      logIfEnabled(
        messagingOptions,
        "client",
        id,
        `response`,
        JSON.stringify(event.data).substring(0, 200)
      );
    });
    return {
      localStorage: {
        setItem: (key, value) => callerWithOptions("localStorage.setItem" /* LocalStorage_SetItem */, key, value),
        getItem: (key) => callerWithOptions("localStorage.getItem" /* LocalStorage_GetItem */, key),
        removeItem: (key) => callerWithOptions("localStorage.removeItem" /* LocalStorage_RemoveItem */, key),
        clear: () => callerWithOptions("localStorage.clear" /* LocalStorage_Clear */),
        key: (index) => callerWithOptions("localStorage.key" /* LocalStorage_Key */, index)
      },
      indexedDBKeyval: {
        set: (key, value) => callerWithOptions("indexDBKeyval.set" /* indexDBKeyval_Set */, key, value),
        get: (key) => callerWithOptions("indexDBKeyval.get" /* indexDBKeyval_Get */, key),
        del: (key) => callerWithOptions("indexDBKeyval.del" /* indexDBKeyval_Del */, key)
      }
    };
  }
  function createIframePostMessage(iframeOptions) {
    const iframe = "src" in iframeOptions ? getOrCreateIframe(iframeOptions.src) : getExistingIframe(iframeOptions.id);
    const { contentWindow } = iframe;
    if (!contentWindow) {
      throw new Error("Injected iframe is missing a contentWindow reference.");
    }
    const readiness = waitForHubReady(iframe);
    return {
      postMessage: contentWindow.postMessage.bind(contentWindow),
      ready: readiness.ready,
      getWaitDurationMs: readiness.getWaitDurationMs,
      isReady: readiness.isReady
    };
  }
  function getOrCreateIframe(iframeSrc) {
    if (typeof document === "undefined") {
      throw new Error("Cannot inject iframe: document is not available.");
    }
    const doc = document;
    const existing = doc.getElementById(DEFAULT_IFRAME_ID);
    if (existing && !(existing instanceof HTMLIFrameElement)) {
      throw new Error(
        `Element with id "${DEFAULT_IFRAME_ID}" already exists and is not an iframe.`
      );
    }
    const iframe = existing ?? doc.createElement("iframe");
    iframe.id = DEFAULT_IFRAME_ID;
    iframe.src = iframeSrc;
    if (!existing) {
      hideIframe(iframe);
      if (!doc.body) {
        throw new Error("Cannot inject iframe: document.body is not available.");
      }
      doc.body.appendChild(iframe);
    }
    return iframe;
  }
  function getExistingIframe(iframeId) {
    if (typeof document === "undefined") {
      throw new Error("Cannot access iframe: document is not available.");
    }
    const existing = document.getElementById(iframeId);
    if (!existing) {
      throw new Error(`Iframe with id "${iframeId}" was not found.`);
    }
    if (!(existing instanceof HTMLIFrameElement)) {
      throw new Error(`Element with id "${iframeId}" is not an iframe.`);
    }
    return existing;
  }
  function hideIframe(iframe) {
    iframe.style.position = "absolute";
    iframe.style.width = "0";
    iframe.style.height = "0";
    iframe.style.border = "0";
    iframe.style.opacity = "0";
    iframe.style.pointerEvents = "none";
  }
  var IFRAME_READY_PING_INTERVAL_MS = 250;
  var iframeReadyStates = /* @__PURE__ */ new WeakMap();
  function waitForHubReady(iframe) {
    const existing = iframeReadyStates.get(iframe);
    if (existing) {
      return {
        ready: existing.ready,
        getWaitDurationMs: () => existing.isReady && existing.resolvedAt ? existing.resolvedAt - existing.startedAt : Date.now() - existing.startedAt,
        isReady: () => existing.isReady
      };
    }
    const startedAt = Date.now();
    let resolvedAt;
    let readyFlag = false;
    const promise = new Promise((resolve) => {
      const { contentWindow } = iframe;
      const handleMessage = (event) => {
        if (event.source !== contentWindow) return;
        if (!isHandshakeMessage(event.data, HANDSHAKE_RESPONSE_TYPE)) return;
        cleanup();
        readyFlag = true;
        resolvedAt = Date.now();
        state.isReady = true;
        state.resolvedAt = resolvedAt;
        resolve();
      };
      const sendPing = () => {
        const message = createHandshakeMessage(HANDSHAKE_REQUEST_TYPE);
        contentWindow.postMessage(message, "*");
      };
      const intervalId = window.setInterval(
        sendPing,
        IFRAME_READY_PING_INTERVAL_MS
      );
      const cleanup = () => {
        window.removeEventListener("message", handleMessage);
        window.clearInterval(intervalId);
      };
      window.addEventListener("message", handleMessage);
      sendPing();
    });
    const state = {
      ready: promise,
      startedAt,
      isReady: readyFlag
    };
    iframeReadyStates.set(iframe, state);
    return {
      ready: promise,
      getWaitDurationMs: () => state.isReady && state.resolvedAt !== void 0 ? state.resolvedAt - state.startedAt : Date.now() - state.startedAt,
      isReady: () => state.isReady
    };
  }
  return __toCommonJS(index_exports);
})();
