// iframe-shared-storage/dist/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.constructClient = constructClient;
const postmsg_rpc_1 = require("postmsg-rpc");
const hub_1 = require("./hub");
const handshake_1 = require("./utils/handshake");
const log_1 = require("./utils/log");
const await_with_timeout_1 = require("./utils/await-with-timeout");
const DEFAULT_IFRAME_ID = "iframe-storage-hub";
const DEFAULT_INITIALIZATION_TIMEOUT_MS = 1000;
const DEFAULT_METHOD_CALL_TIMEOUT_MS = 1000;
const DEFAULT_METHOD_CALL_RETRIES = 0;
class MethodCallTimeoutError extends Error {
    constructor(method, timeoutMs) {
        super(`Iframe storage hub did not respond to method "${method}". Waited ${timeoutMs}ms.`);
        this.method = method;
        this.timeoutMs = timeoutMs;
        this.name = "IframeStorageMethodTimeoutError";
    }
}
function constructClient({ iframe }) {
    const { messagingOptions, iframeReadyTimeoutMs = DEFAULT_INITIALIZATION_TIMEOUT_MS, methodCallTimeoutMs = DEFAULT_METHOD_CALL_TIMEOUT_MS, methodCallRetries = DEFAULT_METHOD_CALL_RETRIES, } = iframe;
    const { postMessage, ready: iframeReady, getWaitDurationMs: getIframeWaitDurationMs, isReady: isIframeReady, } = createIframePostMessage(iframe);
    const callerOptions = { postMessage };
    // Unified dynamic caller to reduce repetition.
    const callerWithOptions = async (method, ...args) => {
        // log a an intent to call a fn (even if it fails later)
        (0, log_1.logIfEnabled)(messagingOptions, "client", method, "before-postmessage", JSON.stringify(args).substring(0, 200));
        // Ensure the hub is initialized before calling the RPC.
        // We don't fail inside the readiness tracker; instead, we enforce the
        // threshold here based on how long initialization has already taken.
        const performCall = async () => {
            if (!isIframeReady()) {
                const elapsed = getIframeWaitDurationMs();
                const remaining = iframeReadyTimeoutMs - elapsed;
                if (remaining <= 0) {
                    throw new Error(`Iframe storage hub did not initialize within the allowed time before calling method "${method}". Waited ${elapsed}ms.`);
                }
                await (0, await_with_timeout_1.awaitWithTimeout)(iframeReady, remaining, () => {
                    throw new Error(`Iframe storage hub did not initialize within the allowed time before calling method "${method}". Waited ${iframeReadyTimeoutMs}ms.`);
                });
            }
            const rpcPromise = (0, postmsg_rpc_1.caller)(method, callerOptions)(...args, messagingOptions);
            return await (0, await_with_timeout_1.awaitWithTimeout)(rpcPromise, methodCallTimeoutMs, () => {
                throw new MethodCallTimeoutError(method, methodCallTimeoutMs);
            });
        };
        let attempt = 0;
        while (attempt <= methodCallRetries) {
            try {
                return await performCall();
            }
            catch (error) {
                const isTimeoutError = error instanceof MethodCallTimeoutError ||
                    (error instanceof Error &&
                        error.name === "IframeStorageMethodTimeoutError");
                const remainingAttempts = methodCallRetries - attempt;
                if (!isTimeoutError || remainingAttempts <= 0) {
                    throw error;
                }
                (0, log_1.logIfEnabled)(messagingOptions, "client", method, "retry-after-timeout", `attempt=${attempt + 1}`, `remaining_attempts=${remainingAttempts - 1}`);
                attempt += 1;
            }
        }
        throw new Error("Unreachable RPC retry loop exit.");
    };
    // for debug purposes
    addEventListener("message", (event) => {
        const sender = event.data.sender;
        if ("postmsg-rpc/server" !== sender)
            return;
        if (messagingOptions?.enableLog !== "client" &&
            messagingOptions?.enableLog !== "both")
            return;
        const id = event.data?.id || "unknown_id";
        (0, log_1.logIfEnabled)(messagingOptions, "client", id, `response`, JSON.stringify(event.data).substring(0, 200));
    });
    return {
        localStorage: {
            setItem: (key, value) => callerWithOptions(hub_1.ApiMethods.LocalStorage_SetItem, key, value),
            getItem: (key) => callerWithOptions(hub_1.ApiMethods.LocalStorage_GetItem, key),
            removeItem: (key) => callerWithOptions(hub_1.ApiMethods.LocalStorage_RemoveItem, key),
            clear: () => callerWithOptions(hub_1.ApiMethods.LocalStorage_Clear),
            key: (index) => callerWithOptions(hub_1.ApiMethods.LocalStorage_Key, index),
        },
        indexedDBKeyval: {
            set: (key, value) => callerWithOptions(hub_1.ApiMethods.indexDBKeyval_Set, key, value),
            get: (key) => callerWithOptions(hub_1.ApiMethods.indexDBKeyval_Get, key),
            del: (key) => callerWithOptions(hub_1.ApiMethods.indexDBKeyval_Del, key),
        },
    };
}
function createIframePostMessage(iframeOptions) {
    const iframe = "src" in iframeOptions
        ? getOrCreateIframe(iframeOptions.src)
        : getExistingIframe(iframeOptions.id);
    const { contentWindow } = iframe;
    if (!contentWindow) {
        throw new Error("Injected iframe is missing a contentWindow reference.");
    }
    const readiness = waitForHubReady(iframe);
    return {
        postMessage: contentWindow.postMessage.bind(contentWindow),
        ready: readiness.ready,
        getWaitDurationMs: readiness.getWaitDurationMs,
        isReady: readiness.isReady,
    };
}
function getOrCreateIframe(iframeSrc) {
    if (typeof document === "undefined") {
        throw new Error("Cannot inject iframe: document is not available.");
    }
    const doc = document;
    const existing = doc.getElementById(DEFAULT_IFRAME_ID);
    if (existing && !(existing instanceof HTMLIFrameElement)) {
        throw new Error(`Element with id "${DEFAULT_IFRAME_ID}" already exists and is not an iframe.`);
    }
    const iframe = existing ?? doc.createElement("iframe");
    iframe.id = DEFAULT_IFRAME_ID;
    iframe.src = iframeSrc;
    if (!existing) {
        hideIframe(iframe);
        if (!doc.body) {
            throw new Error("Cannot inject iframe: document.body is not available.");
        }
        doc.body.appendChild(iframe);
    }
    return iframe;
}
function getExistingIframe(iframeId) {
    if (typeof document === "undefined") {
        throw new Error("Cannot access iframe: document is not available.");
    }
    const existing = document.getElementById(iframeId);
    if (!existing) {
        throw new Error(`Iframe with id "${iframeId}" was not found.`);
    }
    if (!(existing instanceof HTMLIFrameElement)) {
        throw new Error(`Element with id "${iframeId}" is not an iframe.`);
    }
    return existing;
}
function hideIframe(iframe) {
    iframe.style.position = "absolute";
    iframe.style.width = "0";
    iframe.style.height = "0";
    iframe.style.border = "0";
    iframe.style.opacity = "0";
    iframe.style.pointerEvents = "none";
}
const IFRAME_READY_PING_INTERVAL_MS = 250;
const iframeReadyStates = new WeakMap();
function waitForHubReady(iframe) {
    const existing = iframeReadyStates.get(iframe);
    if (existing) {
        return {
            ready: existing.ready,
            getWaitDurationMs: () => existing.isReady && existing.resolvedAt
                ? existing.resolvedAt - existing.startedAt
                : Date.now() - existing.startedAt,
            isReady: () => existing.isReady,
        };
    }
    const startedAt = Date.now();
    let resolvedAt;
    let readyFlag = false;
    const promise = new Promise((resolve) => {
        const { contentWindow } = iframe;
        const handleMessage = (event) => {
            if (event.source !== contentWindow)
                return;
            if (!(0, handshake_1.isHandshakeMessage)(event.data, handshake_1.HANDSHAKE_RESPONSE_TYPE))
                return;
            cleanup();
            readyFlag = true;
            resolvedAt = Date.now();
            state.isReady = true;
            state.resolvedAt = resolvedAt;
            resolve();
        };
        const sendPing = () => {
            const message = (0, handshake_1.createHandshakeMessage)(handshake_1.HANDSHAKE_REQUEST_TYPE);
            // Use "*" to avoid origin mismatches while the iframe is still running
            // about:blank before the remote hub document takes over.
            contentWindow.postMessage(message, "*");
        };
        const intervalId = window.setInterval(sendPing, IFRAME_READY_PING_INTERVAL_MS);
        const cleanup = () => {
            window.removeEventListener("message", handleMessage);
            window.clearInterval(intervalId);
        };
        window.addEventListener("message", handleMessage);
        sendPing();
    });
    const state = {
        ready: promise,
        startedAt,
        isReady: readyFlag,
    };
    iframeReadyStates.set(iframe, state);
    return {
        ready: promise,
        getWaitDurationMs: () => state.isReady && state.resolvedAt !== undefined
            ? state.resolvedAt - state.startedAt
            : Date.now() - state.startedAt,
        isReady: () => state.isReady,
    };
}
