// iframe-shared-storage/dist/hub.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiMethods = void 0;
exports.initHub = initHub;
const postmsg_rpc_1 = require("postmsg-rpc");
const idb_keyval_1 = require("idb-keyval");
const handshake_1 = require("./utils/handshake");
const log_1 = require("./utils/log");
var ApiMethods;
(function (ApiMethods) {
    ApiMethods["LocalStorage_SetItem"] = "localStorage.setItem";
    ApiMethods["LocalStorage_GetItem"] = "localStorage.getItem";
    ApiMethods["LocalStorage_RemoveItem"] = "localStorage.removeItem";
    ApiMethods["LocalStorage_Clear"] = "localStorage.clear";
    ApiMethods["LocalStorage_Key"] = "localStorage.key";
    ApiMethods["indexDBKeyval_Set"] = "indexDBKeyval.set";
    ApiMethods["indexDBKeyval_Get"] = "indexDBKeyval.get";
    ApiMethods["indexDBKeyval_Del"] = "indexDBKeyval.del";
})(ApiMethods || (exports.ApiMethods = ApiMethods = {}));
function initHub() {
    if (!window?.parent)
        throw new Error("Hub must be run inside an iframe with a parent window.");
    // for debug purposes and readiness handshake
    addEventListener("message", (event) => {
        if (event.source === window.parent &&
            (0, handshake_1.isHandshakeMessage)(event.data, handshake_1.HANDSHAKE_REQUEST_TYPE)) {
            window.parent.postMessage((0, handshake_1.createHandshakeMessage)(handshake_1.HANDSHAKE_RESPONSE_TYPE), event.origin);
            return;
        }
        const messagingOptions = event.data?.args
            ? event.data.args[event.data.args.length - 1]
            : undefined;
        if (messagingOptions?.enableLog !== "hub" &&
            messagingOptions?.enableLog !== "both")
            return;
        const method = event.data?.func || "unknown_method";
        const id = event.data?.id || "unknown_id";
        (0, log_1.logIfEnabled)(messagingOptions, "hub", `${id}:${method}`, "received_message", JSON.stringify(event.data).substring(0, 200));
    });
    const localStorageMethods = {
        [ApiMethods.LocalStorage_SetItem]: (key, value) => localStorage.setItem(key, value),
        [ApiMethods.LocalStorage_GetItem]: (key) => localStorage.getItem(key),
        [ApiMethods.LocalStorage_RemoveItem]: (key) => localStorage.removeItem(key),
        [ApiMethods.LocalStorage_Clear]: () => localStorage.clear(),
        [ApiMethods.LocalStorage_Key]: (index) => localStorage.key(index),
    };
    const indexDBKeyvalMethods = {
        [ApiMethods.indexDBKeyval_Set]: (key, value) => (0, idb_keyval_1.set)(key, value),
        [ApiMethods.indexDBKeyval_Get]: (key) => (0, idb_keyval_1.get)(key),
        [ApiMethods.indexDBKeyval_Del]: (key) => (0, idb_keyval_1.del)(key),
        // Note: idb-keyval does not have clear and key methods, so we skip them
    };
    const hubService = {
        ...localStorageMethods,
        ...indexDBKeyvalMethods,
    };
    for (const [methodName, methodImpl] of Object.entries(hubService)) {
        (0, postmsg_rpc_1.expose)(methodName, async (...args) => {
            // Avoid Array.prototype.at for broader lib compatibility.
            const maybeOptions = args.length
                ? args[args.length - 1]
                : undefined;
            const loggedArgs = maybeOptions ? args.slice(0, args.length - 1) : args;
            (0, log_1.logIfEnabled)(maybeOptions, "hub", methodName, "before_call", JSON.stringify(loggedArgs).substring(0, 200));
            const result = await methodImpl(...args);
            (0, log_1.logIfEnabled)(maybeOptions, "hub", methodName, "after_call", String(result).substring(0, 200));
            return result;
        }, {
            postMessage: window.parent.postMessage.bind(window.parent),
        });
    }
}
